# Onramp Data Engineering Project

### Importing Packages

In [81]:
import spotipy
from spotipy.oauth2 import SpotifyOAuth
from spotipy.oauth2 import SpotifyClientCredentials
import pandas as pd
import pprint

### Using spofity app creditentials storted in system setting

In [4]:
auth_manager = SpotifyClientCredentials()
sp = spotipy.Spotify(auth_manager=auth_manager)

## Section 1: Requesting the data

### Creating a list of favorite artists

In [11]:
search_artists = ["Wild Child", "Tessa Violet", "Halsey", "OK Go",
                  "Regina Spektor", "Modest Mouse", "The Killers",
                  "the GazettE", "Dolly Parton", "Ed Sheeran",
                  "The Vaccines", "Bloc Party", "David Bowie",
                  "Creedence Clearwater Revival", "Miike Snow",
                  "Adele", "Grace VanderWaal", "Simple Plan",
                  "blink-182", "MIYAVI"]

### Creating a list of columns for each of the tables

In [ ]:
# df_artist
artist_table_columns = ['artist_id', 'artist_name', 'external_url', 
                        'genre', 'image_url', 'followers', 'popularity', 
                        'type', 'artist_uri']

# df_album
album_table_columns = ['album_id', 'album_name', 'external_url', 
                       'image_url', 'release_date', 'total_tracks', 
                       'type', 'album_uri', 'artist_id']

# df_track
track_table_columns = ['track_id', 'song_name', 'external_url', 'duration_ms', 
                       'explicit', 'disc_number', 'type', 'song_uri', 'album_id']

# df_track_feature
track_feature_table_columns = ['track_id', 'danceability', 'energy', 
                               'instrumentalness', 'liveness', 'loudness', 
                               'speechiness', 'tempo', 'type', 'valence', 'song_uri']

In [10]:
df_artist = pd.DataFrame(columns = artist_table_columns)
df_album = pd.DataFrame(columns = album_table_columns)
df_track = pd.DataFrame(columns = track_table_columns)
df_track_feature = pd.DataFrame(columns = track_feature_table_columns)

## Populating dataframes

### Populating df_artist


In [409]:
def artist_id_list_gen(artist_list):

    for artist in artist_list:

        # Requesting results for each artist in list where type is "artist"
        result = sp.search(q=artist, type="artist")

        # Selecting all data within 'items' in the pulled data
        item = result['artists']['items']
        #print(item)

        # Creating a list of required data to enter into artist_df
        new_row = [item[0]['id'], item[0]['name'], item[0]['external_urls']['spotify'], 
         item[0]['genres'][0], item[0]['images'][0]['url'], 
         item[0]['followers']['total'], item[0]['popularity'],
         item[0]['type'], item[0]['uri']]

        # Inserting pulled data into last row of artist_df
        df_artist.loc[len(df_artist.index)] = new_row

    return df_artist

In [412]:
# clearing out the dataframe so that if the function is run again, duplicates are not added
df_artist = pd.DataFrame(columns = artist_table_columns)

# populating the dataframe
artist_id_list_gen(search_artists)

artist_id                   artist_name  \
0   1xLMexpeeTKQ20SwGMaGSK                    Wild Child   
1   5DD5GZd4ElmQTy9NleMvKJ                  Tessa Violet   
2   26VFTg2z8YR0cCuwLzESi2                        Halsey   
3   3hozsZ9hqNq7CoBGYNlFTz                         OK Go   
4   3z6Gk257P9jNcZbBXJNX5i                Regina Spektor   
5   1yAwtBaoHLEDWAnWR87hBT                  Modest Mouse   
6   0C0XlULifJtAgn6ZNCW2eu                   The Killers   
7   4gRFSxjAcJkb54BDjqo7Bc                   the GazettE   
8   32vWCbZh0xZ4o9gkz4PsEU                  Dolly Parton   
9   6eUKZXaKkcviH0Ku9w2n3V                    Ed Sheeran   
10  0Ak6DLKHtpR6TEEnmcorKA                  The Vaccines   
11  3MM8mtgFzaEJsqbjZBSsHJ                    Bloc Party   
12  0oSGxfWSnnOXhD2fKuz2Gy                   David Bowie   
13  3IYUhFvPQItj6xySrBmZkd  Creedence Clearwater Revival   
14  4l1cKWYW591xnwEGxpUg3J                    Miike Snow   
15  4dpARuHxo51G3z768sgnrY                         Adele   
16  3YdT8QvV4QvY4DfVJhxdcZ              Grace VanderWaal   
17  2p4FqHnazRucYQHyDCdBrJ                   Simple Plan   
18  6FBDaR13swtiWwGhX1WQsP                     blink-182   
19  7sBtBCNVRujQhaHDODkfTN                        MIYAVI   

                                         external_url                genre  \
0   https://open.spotify.com/artist/1xLMexpeeTKQ20...     austin americana   
1   https://open.spotify.com/artist/5DD5GZd4ElmQTy...                alt z   
2   https://open.spotify.com/artist/26VFTg2z8YR0cC...            dance pop   
3   https://open.spotify.com/artist/3hozsZ9hqNq7Co...     alternative rock   
4   https://open.spotify.com/artist/3z6Gk257P9jNcZ...            indie pop   
5   https://open.spotify.com/artist/1yAwtBaoHLEDWA...     alternative rock   
6   https://open.spotify.com/artist/0C0XlULifJtAgn...     alternative rock   
7   https://open.spotify.com/artist/4gRFSxjAcJkb54...              j-metal   
8   https://open.spotify.com/artist/32vWCbZh0xZ4o9...  classic country pop   
9   https://open.spotify.com/artist/6eUKZXaKkcviH0...                  pop   
10  https://open.spotify.com/artist/0Ak6DLKHtpR6TE...           indie rock   
11  https://open.spotify.com/artist/3MM8mtgFzaEJsq...    alternative dance   
12  https://open.spotify.com/artist/0oSGxfWSnnOXhD...             art rock   
13  https://open.spotify.com/artist/3IYUhFvPQItj6x...         classic rock   
14  https://open.spotify.com/artist/4l1cKWYW591xnw...    alternative dance   
15  https://open.spotify.com/artist/4dpARuHxo51G3z...         british soul   
16  https://open.spotify.com/artist/3YdT8QvV4QvY4D...        post-teen pop   
17  https://open.spotify.com/artist/2p4FqHnazRucYQ...    canadian pop punk   
18  https://open.spotify.com/artist/6FBDaR13swtiWw...             pop punk   
19  https://open.spotify.com/artist/7sBtBCNVRujQha...           visual kei   

                                            image_url  followers popularity  \
0   https://i.scdn.co/image/ab6761610000e5eb4d4c9f...     144651         51   
1   https://i.scdn.co/image/ab6761610000e5eb7cc0d7...     647326         57   
2   https://i.scdn.co/image/ab6761610000e5ebd707e1...   18365331         83   
3   https://i.scdn.co/image/0cf0e41aef321054144f04...     483413         56   
4   https://i.scdn.co/image/ab6761610000e5eba36a9f...    1293997         62   
5   https://i.scdn.co/image/ab6761610000e5eb3ff03c...    1309151         63   
6   https://i.scdn.co/image/ab6761610000e5eb207b21...    6648995         77   
7   https://i.scdn.co/image/ab6761610000e5eb19fca7...     287470         48   
8   https://i.scdn.co/image/ab6761610000e5eb20ee5d...    1896365         70   
9   https://i.scdn.co/image/ab6761610000e5eb12a2ef...  102334970         91   
10  https://i.scdn.co/image/ab6761610000e5eb015144...     717042         59   
11  https://i.scdn.co/image/ab6761610000e5ebaedaa0...    1036149         60   
12  https://i.scdn.co/image/ab6761610000e5ebb78f77...    8489243         77   
13  https:/

In [413]:
df_artist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20 entries, 0 to 19
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   artist_id     20 non-null     object
 1   artist_name   20 non-null     object
 2   external_url  20 non-null     object
 3   genre         20 non-null     object
 4   image_url     20 non-null     object
 5   followers     20 non-null     object
 6   popularity    20 non-null     object
 7   type          20 non-null     object
 8   artist_uri    20 non-null     object
dtypes: object(9)
memory usage: 1.6+ KB


#### Remarks on df_artist
All of the artists that I selected are in the dataframe. Additionally none of the values are nulls or duplicates. I can also confirm that only the first genre was selected, not the entire list for each artist.

## Populating album dataframe
To request data for the albums I created a list of all of the artist's ids and I will iterate over those to search for all albums. For the results of each album, I will go through each album, adding the required information to the album dataframe, one row at a time.

In [434]:
artist_ids = df_artist["artist_id"].to_list()

def pull_artist_albums(artist_ids):

    for artist_id in artist_ids:
        # Requesting results for each artist_id in list where type is "album"
        results = sp.artist_albums(artist_id, album_type='album')

        # Selecting all data within 'items' in the pulled data
        items = results['items']

        # Pulling all of the albums from the artist
        for i in range(len(items)):
            new_row = [items[i]['id'], items[i]['name'],items[i]['external_urls'],
                    items[i]['images'][0]['url'], items[i]['release_date'],
                    items[i]['total_tracks'], items[i]['type'],
                    items[i]['uri'], artist_id]

            # Inserting pulled data into last row of album_df
            df_album.loc[len(df_album.index)] = new_row




In [435]:
df_album = pd.DataFrame(columns = album_table_columns)
pull_artist_albums(artist_ids)

Looking at the first five rows of the album dataframe to insure that they are in the same format and data type as expected.

In [437]:
df_album.head()

album_id     album_name  \
0  1RYP4jpaZn8uKR2HLSPLWm   Expectations   
1  5dab4s1UrwHFqLg3tEDnsO   Expectations   
2  13bRrqlPDErhpXDUxYQCUU   Expectations   
3  55ld4JYzuogsLO2utQ1VHk          Fools   
4  7syMxc0OltBTOi0bzUdsyL  The Runaround   

                                        external_url  \
0  {'spotify': 'https://open.spotify.com/album/1R...   
1  {'spotify': 'https://open.spotify.com/album/5d...   
2  {'spotify': 'https://open.spotify.com/album/13...   
3  {'spotify': 'https://open.spotify.com/album/55...   
4  {'spotify': 'https://open.spotify.com/album/7s...   

                                           image_url release_date  \
0  https://i.scdn.co/image/ab67616d0000b27305c904...   2018-02-09   
1  https://i.scdn.co/image/ab67616d0000b273091aed...   2018-02-09   
2  https://i.scdn.co/image/ab67616d0000b273dd907e...   2018-02-09   
3  https://i.scdn.co/image/ab67616d0000b273a7784a...   2015-10-02   
4  https://i.scdn.co/image/ab67616d0000b273c4cc3a...   2015-05-19   

  total_tracks   type                             album_uri  \
0           12  album  spotify:album:1RYP4jpaZn8uKR2HLSPLWm   
1           12  album  spotify:album:5dab4s1UrwHFqLg3tEDnsO   
2           12  album  spotify:album:13bRrqlPDErhpXDUxYQCUU   
3           12  album  spotify:album:55ld4JYzuogsLO2utQ1VHk   
4           11  album  spotify:album:7syMxc0OltBTOi0bzUdsyL   

                artist_id  
0  1xLMexpeeTKQ20SwGMaGSK  
1  1xLMexpeeTKQ20SwGMaGSK  
2  1xLMexpeeTKQ20SwGMaGSK  
3  1xLMexpeeTKQ20SwGMaGSK  
4  1xLMexpeeTKQ20SwGMaGSK

In [438]:
df_album.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 304 entries, 0 to 303
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   album_id      304 non-null    object
 1   album_name    304 non-null    object
 2   external_url  304 non-null    object
 3   image_url     304 non-null    object
 4   release_date  304 non-null    object
 5   total_tracks  304 non-null    object
 6   type          304 non-null    object
 7   album_uri     304 non-null    object
 8   artist_id     304 non-null    object
dtypes: object(9)
memory usage: 23.8+ KB


There are 304 different albums from the 20 artists. A possible issue can be seen from just looking at the head of the dataframe. There are three entries for the album "Expectations" by Wild Child. They all have different ids, but a quick look into the raw data shows the reason these were given different ids was due to the fact they were released in different regions.
</br>
</br>
This also brings up a good point, that there are other possible "duplicates" such as remixes, that may also lie in the data. This will need to be addressed further in Section 2: Data Cleaning.


In [178]:
artist_ids = df_artist["artist_id"].to_list()
#artist_ids = ['1xLMexpeeTKQ20SwGMaGSK']
def pull_artist_albums(artist_ids):
    for artist_id in artist_ids:
        # Requesting results for each artist_id in list where type is "album"
        results = sp.artist_albums(artist_id, album_type='album')
        items = results['items']
        
        # Pulling all of the albums from the artist
       
        for i in range(len(items)):
            # pprint.pprint(items[i])
            new_row = [items[i]['id'], items[i]['name'],items[i]['external_urls']['spotify'],
                    items[i]['images'][0]['url'], items[i]['release_date'],
                    items[i]['total_tracks'], items[i]['type'],
                    items[i]['uri'], artist_id]

            # Inserting pulled data into last row of album_df
            df_album.loc[len(df_album.index)] = new_row

    return df_album


In [179]:
df_album = pd.DataFrame(columns = album_table_columns)
item = pull_artist_albums(artist_ids)
print(df_album)

                   album_id                             album_name  \
0    1RYP4jpaZn8uKR2HLSPLWm                           Expectations   
1    5dab4s1UrwHFqLg3tEDnsO                           Expectations   
2    13bRrqlPDErhpXDUxYQCUU                           Expectations   
3    55ld4JYzuogsLO2utQ1VHk                                  Fools   
4    7syMxc0OltBTOi0bzUdsyL                          The Runaround   
..                      ...                                    ...   
299  6RkfAsqtacwcMlVXUNyEMS   雅 -THIS IZ THE JAPANESE KABUKI ROCK-   
300  7ueVA4xmXnZcgnQNP1Pqf9  7 SAMURAI SESSIONS -We're KAVKI BOIZ-   
301  7qGwHJ4LlCfNiepTFOFoAN                        【雅-みやびうた-歌】~独奏~   
302  0rdMRdjJtYZfxVi8DEMKiZ                               MYV☆POPS   
303  75OnTvYZ4pqKD5ezgFOpGO                        【雅-miyavizm-主義】   

                                          external_url  \
0    https://open.spotify.com/album/1RYP4jpaZn8uKR2...   
1    https://open.spotify.com/album/5dab4s1

# Populating df_track
To retrieve all of the tracks from the albums in df_albums, a list of the album ids was created and the album tracks function from the spotipy package was used. For each album, each result was iterated over to pull all of the necessary track information. The information was stored in the track dataframe.

In [443]:
album_ids = df_album["album_id"].to_list()


def get_tracks(album_ids):
    """ This function searches for the artists name given that
    was given as an argument and returns the first result. If there
    are no result, nothing is returned"""
    for album_id in album_ids:
        results = sp.album_tracks(album_id)
        # Selecting all data within 'items' in the pulled data
        item = results['items']
        
        for i in range(len(item)):
            new_row = [item[i]['uri'], item[i]['name'], item[i]['external_urls']['spotify'],
            item[i]['duration_ms'], item[i]['explicit'], item[i]['disc_number'], item[i]['type'],
            item[i]['uri'], album_id]
            

            # Inserting pulled data into last row of track_df
            df_track.loc[len(df_track.index)] = new_row


In [444]:
df_track = pd.DataFrame(columns = track_table_columns)
get_tracks(album_ids)

In [440]:
df_track.head()

track_id      song_name  \
0  spotify:track:3y5gxbKqif7iw5brilMqEi           Alex   
1  spotify:track:7EgxWK4LlN2t73xmLnyYOK      Eggshells   
2  spotify:track:21gALdjlanjYvfJs6EfvC8   Back & Forth   
3  spotify:track:4dwYda1scNOJxxEPiq6Yg2  Think It Over   
4  spotify:track:3vuwQU2QLKueuzXFzknXm7      Follow Me   

                                        external_url duration_ms explicit  \
0  https://open.spotify.com/track/3y5gxbKqif7iw5b...      176026    False   
1  https://open.spotify.com/track/7EgxWK4LlN2t73x...      309906    False   
2  https://open.spotify.com/track/21gALdjlanjYvfJ...      211360     True   
3  https://open.spotify.com/track/4dwYda1scNOJxxE...      234880    False   
4  https://open.spotify.com/track/3vuwQU2QLKueuzX...      211120    False   

  disc_number   type                              song_uri  \
0           1  track  spotify:track:3y5gxbKqif7iw5brilMqEi   
1           1  track  spotify:track:7EgxWK4LlN2t73xmLnyYOK   
2           1  track  spotify:track:21gALdjlanjYvfJs6EfvC8   
3           1  track  spotify:track:4dwYda1scNOJxxEPiq6Yg2   
4           1  track  spotify:track:3vuwQU2QLKueuzXFzknXm7   

                 album_id  
0  1RYP4jpaZn8uKR2HLSPLWm  
1  1RYP4jpaZn8uKR2HLSPLWm  
2  1RYP4jpaZn8uKR2HLSPLWm  
3  1RYP4jpaZn8uKR2HLSPLWm  
4  1RYP4jpaZn8uKR2HLSPLWm

In [441]:
df_track.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4645 entries, 0 to 4644
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   track_id      4645 non-null   object
 1   song_name     4645 non-null   object
 2   external_url  4645 non-null   object
 3   duration_ms   4645 non-null   object
 4   explicit      4645 non-null   object
 5   disc_number   4645 non-null   object
 6   type          4645 non-null   object
 7   song_uri      4645 non-null   object
 8   album_id      4645 non-null   object
dtypes: object(9)
memory usage: 362.9+ KB


All of the data is in the format that was expected. Furthermore, there is well over the 1000 tracks required. Some investigation will still be needed to check for duplicates.

# Populating df_track_features

In [405]:
track_ids = df_track["track_id"].tolist()

def get_track_info(track_ids):

    for track_id in track_ids:

        result = sp.audio_features(track_id)[0]

        try:
            new_row = [track_id, result['danceability'], result['energy'], result['instrumentalness'], 
                   result['liveness'], result['loudness'], result['speechiness'], result['tempo'], 
                   result['type'], result['valence'], result['uri']]
        except:
            print("There was an issue with a track.")
        df_track_feature.loc[len(df_track_feature.index)] = new_row
        
    return df_track_feature

In [406]:
df_track_feature = pd.DataFrame(columns = track_feature_table_columns)
get_track_info(track_ids)

print(df_track_feature)

There was an issue with a track.
                                  track_id  danceability  energy  \
0     spotify:track:4ktJUA1L4JUwyctMoBQo59         0.470   0.729   
1     spotify:track:6xdaai8nQ2BOT31NtPKXmb         0.407   0.536   
2     spotify:track:3Boa4eIMpJtsSOBkHRdGfh         0.310   0.716   
3     spotify:track:3WHxMUni9hOHbjmUajTkYT         0.650   0.362   
4     spotify:track:4XaDtgAsBJOPn0pYgX8D2Z         0.441   0.819   
...                                    ...           ...     ...   
2844  spotify:track:5zSubCqPi9qyS6zldc1fxa         0.492   0.883   
2845  spotify:track:6KFnxgehoXOtEgJicA1Nqm         0.525   0.974   
2846  spotify:track:0BUE5xWV42yLSuoam0WO6Q         0.484   0.913   
2847  spotify:track:3yK8nkmSan8NE6tXf2v2yy         0.445   0.952   
2848  spotify:track:4jXMozjGEq5S9kLcvJsdQ7         0.506   0.979   

     instrumentalness  liveness  loudness  speechiness    tempo  \
0                   0    0.0932    -4.404       0.0692  100.997   
1            0.0

# Section 2: Cleaning the Data

## 2.1 Handling Null Values

## 2.2 Handling Duplicates

# Creating the database and loading the data

In [385]:
import sqlite3
from sqlite3 import Error

def commit():
    """
    Commits updates to spotify database
    
    Return:
    None
    """
    print ("commit")
    conn.commit()


def create_connection(db_path):
    """
    - Creates and connects to the spotify database
    - Returns the connection and cursor to spotify database
    """

    # connect to spotify database
    try:
        conn = sqlite3.connect(db_path)
        cur = conn.cursor()
        print("Successfully connected to the database")
    except Exception as e:
        print("Error during connection: ", str(e))
    
    return cur, conn


In [398]:
cur, conn = create_database()

Successfully connected to the database


In [283]:
# Drop tables if they exist

drop_table_artist   = "DROP TABLE IF EXISTS artist"
drop_table_album    = "DROP TABLE IF EXISTS album"
drop_table_track    = "DROP TABLE IF EXISTS track"
drop_table_track_feature  = "DROP TABLE IF EXISTS track_feature"

In [319]:
# Creating tables

artist_table_create = ("""CREATE TABLE IF NOT EXISTS artist (
        artist_id      varchar(50)  PRIMARY KEY
        , artist_name  varchar(255)
        , external_url varchar(100)
        , genre        varchar(100)
        , image_url    varchar(100)
        , followers    int
        , popularity   int
        , type         varchar(50)
        , artist_uri   varchar(100)
        )""")

album_table_create = ("""CREATE TABLE IF NOT EXISTS album (
        album_id       varchar(50)  PRIMARY KEY
        , album_name   varchar(255)
        , external_url varchar(100)
        , image_url    varchar(100)
        , release_date date
        , total_tracks int
        , type         varchar(50)
        , album_uri    varchar(100)
        , artist_id    varchar(50)
        )""")

track_table_create = ("""CREATE TABLE IF NOT EXISTS track (
        track_id       varchar(50)  PRIMARY KEY
        , song_name    varchar(255)
        , external_url varchar(100)
        , duration_ms  int
        , explicit     boolean
        , disc_number  int
        , type         varchar(50)
        , song_uri     varchar(100)
        , album_id     varchar(50)
        )""")

track_feature_table_create = ("""CREATE TABLE IF NOT EXISTS track_feature (
        track_id            varchar(50)  PRIMARY KEY
        , daceablility      double
        , energy            double
        , instrumentalness  double
        , liveness          double
        , loudness          double
        , speechiness       double
        , tempo             double
        , type              varchar(50)
        , valence           double
        , song_uri          varchar(100)
        )""")


In [327]:
# Insert data into tables

artist_table_insert = ("""
    INSERT INTO artist (artist_id, artist_name, external_url,
    genre, image_url, followers, popularity, type, artist_uri)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
    ON CONFLICT DO NOTHING
    """)

album_table_insert = ("""
    INSERT INTO album (album_id, album_name, external_url,
    image_url, release_date, total_tracks, type, album_uri,
    artist_id)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
    ON CONFLICT DO NOTHING
    """)

track_table_insert = ("""
    INSERT INTO track (track_id, song_name, external_url,
    duration_ms, explicit, disc_number, type, song_uri,
    album_id)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
    ON CONFLICT DO NOTHING
    """)

track_feature_table_insert = ("""
    INSERT INTO track_feature (track_id, daceablility, energy,
    instrumentalness, liveness, loudness, speechiness, tempo,
    type, valence, song_uri)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
    ON CONFLICT DO NOTHING
    """)

In [392]:
# QUERY LISTS

create_table_queries = [artist_table_create, album_table_create, track_table_create, track_feature_table_create]
drop_table_queries = [drop_table_artist, drop_table_album, drop_table_track, drop_table_track_feature]
table_insert = [(df_artist, 'artist'), (df_album, 'album'), (df_track, 'track')]

In [387]:
def create_tables(conn):
    """
    Create a new tables from the create_tables_queries list
    If an error occurs, the error statement is printed
    
    Parameters:
    conn - connection to the sqlite database
    
    Returns:
    None
    """
    database = r"C:\Users\sbsch\Desktop\spotify.db"
    
    create_connection(database)
    
    for query in create_table_queries:
        try:
            cur = conn.cursor()
            cur.execute(query)
            conn.commit()
        except Error as e:
            print(e)

In [388]:
create_tables(conn)

Successfully connected to the database


In [335]:
def drop_tables(conn):
    """
    Drops tables if the exist from the drop_tables_queries list
    If an error occurs, the error statement is printed
    
    Parameters:
    conn - connection to the sqlite database
    
    Returns:
    None
    """
    for query in drop_table_queries:
        try:
            cur = conn.cursor()
            cur.execute(query)
            conn.commit()
        except Error as e:
            print(e)

In [336]:
drop_tables(conn)

All tables have successfully been droped.
All tables have successfully been droped.
All tables have successfully been droped.
All tables have successfully been droped.


In [400]:
def load_tables(conn):
    """
    Create a new tables from the create_tables_queries list
    If an error occurs, the error statement is printed
    
    Parameters:
    conn - connection to the sqlite database
    
    Returns:
    None
    """
    #artist_data = df_artist.tolist
    for df,table in table_insert:
        try:
            cur = conn.cursor()
            connn = sqlite3.connect('spotify.db')
            #cur.execute(artist_table_insert, artist_data)
            df.to_sql(table, connn, if_exists='replace', index=False)
            conn.commit()
        except Error as e:
            print(e)
    

In [401]:
load_tables(conn)



In [ ]:
    song_data = df[['song_id', 'title', 'artist_id', 'year', 'duration']].values[0].tolist()
    cur.execute(song_table_insert, song_data)